### PR TITLE
Add more units tests to attest edge cases involving errors - Part 1

### PR DESCRIPTION
**Pull request by c410-f3r**
Changes are becoming quite large so this PR only brings a subset of all required additional unit tests for an easier review. More tests will be added afterwards.

Code should be read commit-by-commit for a better understanding of what is going on.

```
    Add more units tests to attest edge cases involving errors
    
    Beyond the addition of more tests, also moves very large common errors into unified modules.
```

```
    Use built-in overflow errors
    
    Substrate already provides built-in errors for arithmetic errors so custom error enum variants are unnecessary
```

```
    Re-arrange modules, exports and cfg flags
    
    `cfg` flags for modules can be placed inside the module itself enhacing overall readability.
    Also groups scattered exports and modules declarations.
 ```

### CODE REVIEW

**Comment by c410-f3r** [comment]
At least locally, tests are very flaky and I am not sure what is causing this intermittent behaviour.

File: pallets/loans/src/lib.rs [hunk 1]
```diff
@@ -1332,11 +1306,11 @@ impl<T: Config> Pallet<T> {
     pub fn calc_collateral_amount(
         underlying_amount: u128,
         exchange_rate: Rate,
-    ) -> result::Result<Balance, Error<T>> {
+    ) -> result::Result<Balance, DispatchError> {
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:1309]: result:: is not necessary here

File: pallets/loans/src/lib.rs [hunk 2]
```diff
@@ -45,18 +47,15 @@ use sp_runtime::{
...
 
-#[cfg(test)]
 mod mock;
 mod rate;
-#[cfg(test)]
 mod tests;
+pub mod weights;
+
+pub(crate) const OVERFLOW: DispatchError = DispatchError::Arithmetic(ArithmeticError::Overflow);
+pub(crate) const UNDERFLOW: DispatchError = DispatchError::Arithmetic(ArithmeticError::Overflow);
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:58]: we can remove `pub(crate)` if this will only be used by this pallet, if we would like to keep it we can move it to a common place

also this should be
```
 pub(crate) const UNDERFLOW: DispatchError = DispatchError::Arithmetic(ArithmeticError::Underflow);
 ```
right?

File: pallets/loans/src/lib.rs [hunk 3]
```diff
@@ -1010,7 +990,7 @@ impl<T: Config> Pallet<T> {
     pub fn borrow_balance_stored_with_snapshot(
         currency_id: &CurrencyId,
         snapshot: BorrowSnapshot,
-    ) -> result::Result<Balance, Error<T>> {
+    ) -> result::Result<Balance, DispatchError> {
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:993]: I forgot why do we need to type `result::Result` instead of `Result`

File: pallets/loans/src/lib.rs (hunk 2 above)
Comment (yz89) [reply pallets/loans/src/lib.rs:58]: `ArithmeticError` already implemented the `From` for `DispatchError`. I don't think it's necessary to type the const one.
https://github.com/paritytech/substrate/blob/master/primitives/runtime/src/lib.rs#L593

Comment (yz89) [reply pallets/loans/src/lib.rs:58]: We can use `ArithmeticError` directly like this. https://github.com/AcalaNetwork/Acala/blob/master/modules/loans/src/lib.rs#L325

File: pallets/loans/src/lib.rs (hunk 3 above)
Comment (yz89) [reply pallets/loans/src/lib.rs:993]: Because we can not use `std::Result` in wasm runtime. 

`sp_std::result` is provided by substrate.

Comment (GopherJ) [reply pallets/loans/src/lib.rs:993]: then we can import sp_std::result::Result maybe

Comment (yz89) [reply pallets/loans/src/lib.rs:993]: That makes sense.

File: pallets/loans/src/tests/liquidate_borrow.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    ExtBuilder::default().build().execute_with(|| {
+        initial_setup();
+        alice_borrows_100_ksm();
+        MOCK_PRICE_FEEDER::set_price(KSM, Rate::from_float(2000.0));
+        LiquidationIncentive::<Runtime>::insert(KSM, Rate::from_float(200.0));
+        assert_noop!(
+            Loans::liquidate_borrow(Origin::signed(BOB), ALICE, KSM, million_dollar(50), DOT),
+            Error::<Runtime>::RepayValueGreaterThanCollateral
+        );
+    })
```
Comment (yz89) [inline pallets/loans/src/tests/liquidate_borrow.rs:33]: MOCK_PRICE_FEEDER::reset();

File: pallets/loans/src/tests/liquidate_borrow.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    ExtBuilder::default().build().execute_with(|| {
+        initial_setup();
+        alice_borrows_100_ksm();
+        MOCK_PRICE_FEEDER::set_price(KSM, 20.into());
+        assert_noop!(
+            Loans::liquidate_borrow(Origin::signed(BOB), ALICE, KSM, million_dollar(51), DOT),
+            Error::<Runtime>::RepayAmountExceedsCloseFactor
+        );
+    })
+}
```
Comment (yz89) [inline pallets/loans/src/tests/liquidate_borrow.rs:101]: MOCK_PRICE_FEEDER::reset();

File: pallets/loans/src/lib.rs [hunk 6]
```diff
@@ -34,9 +34,11 @@ use frame_support::{
 };
 use frame_system::pallet_prelude::*;
 use orml_traits::{MultiCurrency, MultiCurrencyExtended};
+pub use pallet::*;
```
Comment (ayushmishra2005) [inline pallets/loans/src/lib.rs:37]: Why do we need this?

Comment (kaichaosun) [reply pallets/loans/src/lib.rs:37]: Runtime needs to find the data types defined in this module.

File: pallets/loans/src/tests/liquidate_borrow.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        initial_setup();
+        assert_noop!(
+            Loans::liquidate_borrow(Origin::signed(BOB), ALICE, KSM, 0, DOT),
+            Error::<Runtime>::NoBorrowBalance
+        );
+    })
+}
+
+#[test]
+fn collateral_value_must_be_less_than_liquidation_value() {
```
Comment (kaichaosun) [inline pallets/loans/src/tests/liquidate_borrow.rs:23]: less -> bigger

Comment (c410-f3r) [reply pallets/loans/src/tests/liquidate_borrow.rs]: Done

File: pallets/loans/src/tests/liquidate_borrow.rs (hunk 5 above)
Comment (c410-f3r) [reply pallets/loans/src/tests/liquidate_borrow.rs:100]: Included

File: pallets/loans/src/lib.rs (hunk 3 above)
Comment (c410-f3r) [reply pallets/loans/src/lib.rs]: Modified

File: pallets/loans/src/lib.rs (hunk 2 above)
Comment (c410-f3r) [reply pallets/loans/src/lib.rs]: Removed in favor of local declarations

**Comment by yz89** [comment]
> At least locally, tests are very flaky and I am not sure what is causing this intermittent behavior.
> 
> EDIT: Probably because of modified storage parameters. I will investigate tomorrow.

It mostly caused by `MOCK_PRICE_FEEDER` data race. Let's consider how to use `MOCK_PRICE_FEEDER` in a multi-thread environment.

File: pallets/loans/src/tests.rs [hunk 8]
```diff
@@ -934,3 +898,14 @@ fn get_price_works() {
...
+// Groups all tests that conflict internal storage due to concurrent reads and writes
+// or in other words, groups all flaky concurrent tests.
+#[test]
+fn sequential_tests() {
+    total_collateral_asset_value_works();
+
+    liquidate_borrow::collateral_value_must_be_greater_than_liquidation_value();
+    liquidate_borrow::full_workflow_works_as_expected();
+    liquidate_borrow::liquidator_can_not_repay_more_than_the_close_factor_pct_multiplier();
+}
```
Comment (c410-f3r) [inline pallets/loans/src/tests.rs:911]: `cargo test` spawns `N`, where `N = number of host cores`, jobs so parallel by default and by design.
To avoid having to create a sequential entry-point for tests, it is also possible to always run `cargo test -- --test-threads 1`

Comment (yz89) [reply pallets/loans/src/tests.rs:911]: That makes sense. Maybe this will fix the intermittent problems in Test Coverage. @GopherJ

**Review by yz89 — APPROVED** [review]
